examples/sat/hillclimber.ipynb

```diff
@@ -0,0 +1,1097 @@
+       "    \n",
+       "    Args:\n",
+       "        iterable: iterable to find the maximum in\n",
+       "        default: default value to return if iterable is empty.\n",
+       "        key: function that maps iterable elements to values that are comparable. By default, use the iterable elements.\n",
+       "        error: upper bound on the failure probability of the quantum algorithm.\n",
+       "        stats: object that keeps track of statistics.\n",
+       "    \n",
+       "    Raises:\n",
+       "        ValueError: Raised when the failure rate `error` is not provided and statistics cannot be calculated.\n",
+       "        ValueError: Raised when iterable is an empty sequence and no default is provided.\n",
+       "    \n",
+       "    Returns:\n",
+       "        the desired maximum element\n",
+       "\u001B[0;31mCall docstring:\u001B[0m\n",
+       "    Call self as a function."
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "from qubrabench.algorithms.max import max as qmax\n",
+    "\n",
+    "%pdoc qmax"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following example will use the max function documented above. It may seem a bit confusing, that the values in the array (iterable) should be themselves arrays containing only a single integer, however this will be used here to demonstrate the use of the key function. We provide the max function with the described array and intend to receive the array containing the highest integer. The default value is set to [-1] here, such a value would indicate an empty iterable being passed to the function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[91]"
+      ]
+     },
+     "execution_count": null,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from qubrabench.algorithms.max import max\n",
+    "\n",
+    "max([[14], [2], [30], [7], [91]], default=[-1], key=lambda x: x[0])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that we have a basic understanding of the search and max functions, we can have a closer look at the MAX-k-SAT problem, which we intend to solve using our hillclimber algorithm."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Problem: MAX-k-SAT\n",
+    "\n",
+    "Max-k-SAT is a combinatorial optimization problem that given a list of clauses $(C_{i})^{p}_{i=1}$, each a disjunction of at most $k$ literals, and a set of weights $(w_{i})^{p}_{i=1}$, asks us to maximize the weight of the satisfied clauses,\n",
+    "$$\\varphi(y) := \\sum ^{p} _{i=1} w_{i}C_{i}(y),$$\n",
+    "over all assignments $y \\in \\{0, 1\\}^{q}$ of the variables. This problem is NP-hard for $k ≥ 2$.\n",
```

Comment:
Isn't it NP-hard for $`k \geq 3`$ ?

Reply:
You're confusing this problem with SAT (a decision problem). 2-SAT is P, 3-SAT is NP-complete.

But MAX-k-SAT is a different problem, which is NP-hard even for k=2 (see [this](https://en.wikipedia.org/wiki/2-satisfiability#Maximum-2-satisfiability))